### PR TITLE
Final tests for diagonal

### PR DESCRIPTION
**Pull request by kshyatt**
This should get us to 100% coverage again for this file

### CODE REVIEW

**Comment by kshyatt** [comment]
Whoops I got carried away and now there are a few other small things in here

File: test/linalg/diagonal.jl [hunk 1]
```diff
@@ -91,6 +91,8 @@ srand(1)
                     Uc = ctranspose(U)
                     target = scale!(Uc,inv.(D.diag))
                     @test A_rdiv_B!(Uc,D) ≈ target atol=2n^3*eps(relty)*(1+(elty<:Complex))
+                    @test_throws DimensionMismatch A_rdiv_B!(eye(elty, n-1),D)
```
Comment (tkelman) [inline test/linalg/diagonal.jl]: space after the comma before D

File: test/linalg/diagonal.jl [hunk 2]
```diff
@@ -158,6 +160,14 @@ srand(1)
         @test (r = full(D) * U   ; A_mul_B!(UU, D, U) ≈ r ≈ UU)
         @test (r = full(D)' * U  ; Ac_mul_B!(UU, D, U) ≈ r ≈ UU)
         @test (r = full(D).' * U ; At_mul_B!(UU, D, U) ≈ r ≈ UU)
+
+        VV = full(D)
+        DD = copy(D)
+        @test (r = VV * full(D)   ; full(A_mul_B!(VV, DD)) ≈ r ≈ full(D)*full(D))
```
Comment (tkelman) [inline test/linalg/diagonal.jl]: why does the `r = VV * full(D)` need to be inside the `@test` instead of on its own line?

Comment (kshyatt) [inline test/linalg/diagonal.jl]: Consistence with the tests right above. I can modify all of them if you'd like?

Comment (tkelman) [inline test/linalg/diagonal.jl]: the spacing looks a bit strange here, didn't look higher in the file to see if that's copied

File: test/linalg/diagonal.jl (hunk 1 above)
Comment (kshyatt) [inline test/linalg/diagonal.jl]: I spaced out the commas all around this and set the tolerances as their own variables so we can change them all together (and the wrapping is better)...

File: test/linalg/diagonal.jl (hunk 2 above)
Comment (kshyatt) [inline test/linalg/diagonal.jl]: It was.

Comment (tkelman) [inline test/linalg/diagonal.jl]: the line below ends up no longer horizontally aligned which makes this formatting look superfluous

File: test/linalg/diagonal.jl [hunk 3]
```diff
@@ -158,6 +162,17 @@ srand(1)
...
+        VV = full(D)
+        DD = copy(D)
+        r  = VV * full(D)
+        @test full(A_mul_B!(VV, DD)) ≈ r ≈ full(D)*full(D)
+        DD = copy(D)
+        r  = VV * (full(D).')
+        @test full(A_mul_Bt!(VV, DD)) ≈ r
+        DD = copy(D)
+        r  = VV * (full(D)')
+        @test full(A_mul_Bc!(VV, DD)) ≈ r
```
Comment (Sacha0) [inline test/linalg/diagonal.jl]: Perhaps disambiguate the `full` calls in this block? That is, perhaps use `Array(...)` or `AbstractArray(...)` as appropriate instead?

Comment (Sacha0) [inline test/linalg/diagonal.jl]: An explanatory comment would be lovely as well :).

File: test/linalg/generic.jl [hunk 4]
```diff
@@ -342,3 +342,6 @@ Base.abs(a::ModInt{n}) where {n} = a
 Base.:<(a::ModInt{n}, b::ModInt{n}) where {n} = a.k < b.k
 
 @test A*(lufact(A, Val{true})\b) == b
+
+@test_throws ErrorException ctranspose(rand(2,2,2,2))
+@test_throws ErrorException transpose(rand(2,2,2,2))
```
Comment (Sacha0) [inline test/linalg/generic.jl]: Are these checking that `[c]transpose` fails for rank greater than two? If so, perhaps an explanatory comment? :)

File: test/linalg/rowvector.jl [hunk 5]
```diff
@@ -268,3 +268,19 @@ end
 @testset "ambiguity between * methods with RowVectors and ConjRowVectors (#20971)" begin
     @test RowVector(ConjArray(ones(4))) * ones(4) == 4
 end
+
+@testset "setindex!/getindex" begin
+    v = [2,3,4]
```
Comment (Sacha0) [inline test/linalg/rowvector.jl]: I have been trying to develop a model of when you do / do not inject spaces but haven't been successful yet ;).

File: test/linalg/symmetric.jl [hunk 6]
```diff
@@ -224,6 +224,18 @@ end
...
+        if eltya <: Real
+            @test asym^2 ≈ Symmetric((a+a')^2)
+            @test asym^-2 ≈ Symmetric((a+a')^-2)
+            @test asym^2.0 ≈ Symmetric((a+a')^2.0)
+            @test asym^-2.0 ≈ Symmetric((a+a')^-2.0)
+        else
+            @test asym^2 ≈ Hermitian((a+a')^2)
+            @test full(asym^-2) ≈ (a+a')^-2
+            @test asym^2.0 ≈ Hermitian((a+a')^2.0)
+            @test full(asym^-2.0) ≈ (a+a')^-2.0
```
Comment (Sacha0) [inline test/linalg/symmetric.jl]: Perhaps replace the `full` calls with `Array`, `AbstractArray`, or whichever unambiguous call is appropriate?

File: test/linalg/rowvector.jl (hunk 5 above)
Comment (kshyatt) [inline test/linalg/rowvector.jl]: Same here lol, the heart cannot know itself

**Comment by kshyatt** [comment]
@Sacha0 better?

**Review by KristofferC — APPROVED** [review]

**Comment by kshyatt** [comment]
Bump for @Sacha0 and @tkelman

**Review by Sacha0 — APPROVED** [review]
lgtm! Thanks @kshyatt! :)

**Comment by kshyatt** [comment]
Woohoo! `diagonal.jl` is 100% covered again :)

**Pull request merged**